extensions/bitwarden/src/components.tsx

```diff
@@ -57,3 +64,445 @@ export function UnlockForm(props: {
+  }
+}
+
+export function LoginForm(props: {
+  bitwardenApi: Bitwarden;
+  sessionToken?: string
+}) {
+  const { sessionToken, bitwardenApi } = props;
+  const [options, setOptions] = useState<PasswordOptions>({
+    len: 15,
+    uppercase: false,
+    lowercase: false,
+    numeric: false,
+    special: false
+  });
+  const [passphraseOptions, setPassphraseOptions] = useState<PassphraseOptions>({
+    wordsLen: 0,
+    separator: "",
+    capitalize: false,
+    numbers: false
+  });
+  const [passOpt, setPassOpt] = useState<string>("none");
+  const [additionalOptions, setAdditionalOptions] = useState<boolean>(false);
+  const [pass, setPass] = useState<string>("");
+  const [uris, setUris] = useState<Uris[]>([]);
+  const [uriCount, setUriCount] = useState<number>(0);
+  const [fieldCount, setFieldCount] = useState<number>(0);
+
+  async function genPassword(opts: PasswordOptions | PassphraseOptions) {
+    setPass(await bitwardenApi.generate(opts));
+  }
+
+  return <Form navigationTitle={"Add Login 🧍"} isLoading={sessionToken == null}
+               actions={
+                 <ActionList submitForm={e => {
+                   // login form type
+                   e.type = 1;
+                   //needed to re-map form values to array
+                   e.uriCount = uriCount;
+                   e.fieldCount = fieldCount;
+                   submitForm(e, bitwardenApi, sessionToken);
+                 }} form={"login"}
+                             bitwardenApi={bitwardenApi}
+                             sessionToken={sessionToken} />
+               }>
+    {sessionToken && <Fragment>
+      <Form.Description text={"Create login"} />
+      <Form.TextField id={"name"} title={"Name"} />
+      <Form.TextField id={"userName"} title={"User Name"} />
+      <Form.TextField id={"otp"} title={"Authenticator key (TOTP)"} />
+      <Form.TextField id={"password"} title={"Password"} value={pass} onChange={setPass} />
+      <Form.Dropdown id={"gen"} title={"Password Options"} onChange={e => {
+        switch (e) {
+          case "password":
+            setPassOpt(e);
+            genPassword(options);
+            break;
+          case "passphrase":
+            setPassOpt(e);
+            genPassword(passphraseOptions);
+            break;
+          default:
+            setPassOpt(e);
+        }
+      }}>
+        <Form.DropdownItem value={"none"} title={"Manual"} />
+        <Form.DropdownItem value={"password"} title={"Password"} />
+        <Form.DropdownItem value={"passphrase"} title={"Passphrase"} />
+      </Form.Dropdown>
+      {passOpt == "password" &&
+        <Fragment key={"passwordOptions"}>
+          <Form.TextField id={"passLen"} title={"Password Length"} onChange={e => {
+            if (options.len != +e) {
+              options.len = +e;
+              setOptions(options);
+              genPassword(options);
+            }
+          }} storeValue={true} />
```

Comment:
i'm using `storeValue` to save user's choice, but it leads to 0-5 re-renders on each `onChange` might i can replace it with `localStorage` also because of async generation sometimes first generated password doesn't use all options.